File: python/flyql/core/expression.py
from dataclasses import dataclass, field
from typing import Any, List, Optional, Tuple
from flyql.core.exceptions import FlyqlError
from flyql.core.constants import VALID_KEY_VALUE_OPERATORS, Operator
from flyql.core.key import Key
from flyql.core.range import Range
from flyql.literal import LiteralKind


@dataclass
class Duration:
    value: int
    unit: str


@dataclass
class FunctionCall:
    name: str
    duration_args: list["Duration"] = field(default_factory=list)
    unit: str = ""
    timezone: str = ""
    parameter_args: list["Parameter"] = field(default_factory=list)


@dataclass
class Parameter:
    name: str
    positional: bool


INT64_MIN = -(2**63)
INT64_MAX = 2**63 - 1


def convert_unquoted_value(
    value: str | int | float,
) -> Tuple[str | int | float, LiteralKind]:
    try:
        int_val = int(str(value))
        if INT64_MIN <= int_val <= INT64_MAX:
            return int_val, LiteralKind.INTEGER
        return int_val, LiteralKind.BIGINT
    except ValueError:
        pass
    try:
        return float(value), LiteralKind.FLOAT
    except ValueError:
        return value, LiteralKind.COLUMN


class Expression:
    def __init__(
        self,
        key: Key,
        operator: str,
        value: "str | int | float | bool | None | FunctionCall | Parameter",
        value_is_string: bool | None,
        range: Optional[Range] = None,
        operator_range: Optional[Range] = None,
        value_range: Optional[Range] = None,
        value_ranges: Optional[List[Range]] = None,
        values: Optional[List[Any]] = None,
        values_type: Optional[str] = None,
        value_type: Optional[LiteralKind] = None,
        values_types: Optional[List[LiteralKind]] = None,
    ) -> None:
        if operator not in VALID_KEY_VALUE_OPERATORS:
            raise FlyqlError(f"invalid operator: {operator}")

        if not key.segments:
            raise FlyqlError("empty key")

        self.key = key
        self.operator = operator
        self.values: Optional[List[Any]] = values
        self.values_type: Optional[str] = values_type
        self.values_types: Optional[List[LiteralKind]] = values_types
        self.range = range if range is not None else Range(0, 0)
        self.operator_range = operator_range
        self.value_range = value_range
        self.value_ranges = value_ranges

        if value_type is not None:
            self.value: Any = value
            self.value_type: Optional[LiteralKind] = value_type
        elif operator in (Operator.IN.value, Operator.NOT_IN.value):
            self.value = None
            self.value_type = None
        elif (
            not value_is_string
            and value is not None
            and not isinstance(value, (FunctionCall, Parameter))
        ):
            self.value, self.value_type = convert_unquoted_value(value)
        else:
            self.value = value
            self.value_type = LiteralKind.STRING

    def __str__(self) -> str:
        if self.operator in (Operator.IN.value, Operator.NOT_IN.value):
            return f"{self.key.raw} {self.operator} [{', '.join(str(v) for v in (self.values or []))}]"
        display_value = "null" if self.value is None else self.value
        return f"{self.key.raw}{self.operator}{display_value}"

File: javascript/src/core/expression.js
import { convertUnquotedValue } from './utils.js'
import { FlyqlError } from './exceptions.js'
import { VALID_KEY_VALUE_OPERATORS, Operator } from './constants.js'
import { Key } from './key.js'
import { ValueType } from '../types.js'

export class Duration {
    constructor(value, unit) {
        this.value = value
        this.unit = unit
    }
}

export class FunctionCall {
    constructor(name, durationArgs = [], unit = '', timezone = '') {
        this.name = name
        this.durationArgs = durationArgs
        this.unit = unit
        this.timezone = timezone
        this.parameterArgs = []
    }
}

export class Parameter {
    constructor(name, positional) {
        this.name = name
        this.positional = positional
    }
}

export class Expression {
    constructor(
        key,
        operator,
        value,
        valueIsString = null,
        values = null,
        valuesType = null,
        valuesTypes = null,
        valueType = undefined,
        range = null,
        operatorRange = null,
        valueRange = null,
        valueRanges = null,
    ) {
        if (operator !== Operator.TRUTHY && !VALID_KEY_VALUE_OPERATORS.includes(operator)) {
            throw new FlyqlError(`invalid operator: ${operator}`)
        }

        if (!key.segments || key.segments.length === 0) {
            throw new FlyqlError('empty key')
        }

        this.key = key
        this.operator = operator
        this.values = values
        this.valuesType = valuesType
        this.valuesTypes = valuesTypes
        this.range = range
        this.operatorRange = operatorRange
        this.valueRange = valueRange
        this.valueRanges = valueRanges

        if (valueType !== undefined) {
            this.value = value
            this.valueType = valueType
        } else if (operator === Operator.TRUTHY) {
            this.value = ''
            this.valueType = ValueType.STRING
        } else if (operator === Operator.IN || operator === Operator.NOT_IN) {
            this.value = ''
            this.valueType = null
        } else if (valueIsString === false) {
            const [convertedValue, detectedType] = convertUnquotedValue(value)
            this.value = convertedValue
            this.valueType = detectedType
        } else if (valueIsString === true) {
            this.value = value
            this.valueType = ValueType.STRING
        } else {
            const [convertedValue, detectedType] = convertUnquotedValue(value)
            this.value = convertedValue
            this.valueType = detectedType
        }
    }

    toString() {
        if (this.operator === Operator.IN || this.operator === Operator.NOT_IN) {
            const valuesStr = this.values ? this.values.join(', ') : ''
            return `${this.key.raw} ${this.operator} [${valuesStr}]`
        }
        return `${this.key.raw}${this.operator}${this.value}`
    }
}
